### PR TITLE
Add type annotations for unit tests

### PR DESCRIPTION
**Pull request by mthuurne**
My hope is that the calls made by the test cases to the code under test will act as a sanity check when adding type annotations to that code.

In any case, the process of adding the annotations to the unit tests led to several code and test cleanups already. I put these in separate commits, for easier reviewing.

### CODE REVIEW

File: pydoctor/epydoc/markup/__init__.py
Comment (graingert) [inline pydoctor/epydoc/markup/__init__.py:113]:
```suggestion
def flatten(stan: object) -> str:
```

Comment (graingert) [reply pydoctor/epydoc/markup/__init__.py:113]: Any is much wider than object

Comment (mthuurne) [reply pydoctor/epydoc/markup/__init__.py:113]: The problem with `object` is that as soon as `twisted.web.template.flattenString()` receives annotations, mypy will flag an error when passing an `object` as the second argument. While `Any` is not very descriptive, it is less fragile.

Comment (graingert) [reply pydoctor/epydoc/markup/__init__.py:113]: That's the advantage of using object here, you get notified of the type change straight away

Comment (graingert) [reply pydoctor/epydoc/markup/__init__.py:113]: You might want to set https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-disallow-any-explicit too

Comment (mthuurne) [reply pydoctor/epydoc/markup/__init__.py:113]: But it won't happen straight away: it will happen on some random PR, which then either has to include a fix for an unrelated issue or be put on hold until a fix has been merged.

Comment (mthuurne) [reply pydoctor/epydoc/markup/__init__.py:113]: pydoctor is almost 10k lines of code and very little of it has type annotations yet. The first priority is to get mypy coverage up to a level where it can detect inconsistencies. Making the ruleset  stricter is a good idea for the future, but doing that right would be counter-productive, in my opinion.

Comment (mthuurne) [reply pydoctor/epydoc/markup/__init__.py]: I think I found a workable compromise: accept only `Tag` objects. That gives mypy a specific type to work with, while we can always wrap strings, sequences etc. in a transparent tag to convert them to a `Tag` object.

File: pydoctor/driver.py
Comment (graingert) [inline pydoctor/driver.py:333]: ?

Comment (graingert) [reply pydoctor/driver.py:333]: ah it's redundant

File: tox.ini
Comment (graingert) [inline tox.ini:61]:
```suggestion
```

I don't think you need this - as it's only imported in `if TYPE_CHECKING:`

**Review by graingert — APPROVED** [review]
double check you need the typing-extensions dep, but otherwise good to go

File: pydoctor/driver.py
Comment (mthuurne) [reply pydoctor/driver.py:333]: Unfortunately GitHub's review system doesn't do a great job presenting details from commit comments: you either have to review individual commits or you'll get the total change set without any context. It would be useful if they had some kind of `git blame` integration to show which commit is responsible for each changed line.

**Comment by mthuurne** [comment]
> double check you need the typing-extensions dep, but otherwise good to go

I couldn't find anything in mypy's docs about providing typing-extensions for us, so I'm going to assume that if they are available, it's by accident rather than by design, in which case we should still be declaring the dependency. Note that the dependency is in the `mypy` tox env only, so it doesn't affect unit/system tests or the wheel.

**Pull request merged**
